codecell/todoApi
# #Todos web app API implementation using mongoDB, express and node js (MEN) stack

**What does this PR do?**
- This is for the review of the todos feature implementation.

**Endpoints created**
- GET '/' --> Home route
- POST '/todos' --> Create a new todo.
- GET '/todos'  --> View all Todos.
- GET '/todos/:id' --> View a Todo with specified ID.
- PATCH  '/todos/:id' --> Update a Todo with specified ID.
- DELETE  '/todos/:id' --> Delete a Todo with specified ID.
-  GET '*' --> Not found (404) route.

**How can this be manually tested?**
- Ensure You have [nodeJs](https://nodejs.org/en/), [MongoDB](https://www.mongodb.com/try/download/community) and [Postman](https://www.postman.com/downloads/)  installed .
- Using your MongoDB compass or Terminal, create a database named **todoDb**.
- Also create a second DB for unit/integration tests named **todoTests**.
- After cloning this repo
- Create a local branch named `ft/todos`.
- Run `git pull origin ft/todos` to pull in all changes from this branch.
- Run `npm install` to install the project dependencies.
- Run `npm run start-dev` to serve the project
- Open postman and in a postman tab, use the following to test the endpoints;
1. ) **Create new Todo route**
- In postman,  ensure  **raw** and **JSON** are selected, make a **POST** request to the endpoint **http://localhost:3000/todos**  with data of the following shape
{
    "priority": 1,
    "description": "Go get this done",
}

- on Sucess, you should get a JSON response with a status code of 201 with some info and data.

2.) **View all todos Route**
make a **GET** request to the endpoint **http://localhost:3000/todos**
- on Sucess, you should get a JSON response with a status code of 200 with some info and data containing all todos created.

3.) **Update  todos Route**
-**Note: ** You need the ID of a todo to be able to update or delete it, using the get request in step 2 above, copy the an ID of any of the todos you created and replace **:id**  in the following route with the todo ID.
**http://localhost:3000/todos/:id**,
Then make a **PATCH** request with any OR all of the attributes of a todo e.g
{
    "priority": 1,
    "description": "Go get this done",
"completed": true
}
- on Sucess, you should get a JSON response with a status code of 200 with some info and data containing the updated todo,
- Ideally, this endpoint should return a status code of 204 but I used 200 to be able to pass a message.

4.) **Delete  todos Route**
- To delete a todo
- Similar to step 3 above, if you still have the tab you used to update the todo in step 3 above, just change the postman REST API method from PATCH to DELETE, and send,
- on Sucess, you should get a JSON response with a status code of 200 with some info about the deleted todo,
- Ideally, this endpoint should  also return a status code of 204 but I used 200 to be able to pass an info.

## UNIT TESTS
- Once you've cloned the repo and ran npm install,
- Ensure the development server is not currently running to avoid duplicate calls.
- Run `npm run test` to run all the unit tests written for end points.
